src/signTxCatalystRegistration.c

```diff
@@ -0,0 +1,576 @@
+	                                   commonTxData->networkId
+	                           );
+	TRACE("Policy: %d", (int) policy);
+	ENSURE_NOT_DENIED(policy);
+
+	{
+		ASSERT(isShelleyAddressType(subctx->stateData.votingRewardsAddressParams.type));
+		uint8_t addressBuffer[MAX_ADDRESS_SIZE];
+		size_t addressSize;
+		addressSize = deriveAddress(
+		                      &subctx->stateData.votingRewardsAddressParams,
+		                      addressBuffer,
+		                      SIZEOF(addressBuffer)
+		              );
+		ASSERT(addressSize > 0);
+		ASSERT(addressSize < BUFFER_SIZE_PARANOIA);
+
+		auxDataHashBuilder_catalystRegistration_addVotingRewardsAddress(
+		        auxDataHashBuilder, addressBuffer, addressSize
+		);
+	}
+
+	{
+		// select UI steps
+		switch (policy) {
+#	define  CASE(POLICY, UI_STEP) case POLICY: {subctx->ui_step=UI_STEP; break;}
+			CASE(POLICY_PROMPT_WARN_UNUSUAL, HANDLE_VOTING_REWARDS_ADDRESS_PARAMS_STEP_WARNING);
+			CASE(POLICY_SHOW_BEFORE_RESPONSE, HANDLE_VOTING_REWARDS_ADDRESS_PARAMS_STEP_DISPLAY_ADDRESS);
+			CASE(POLICY_ALLOW_WITHOUT_PROMPT, HANDLE_VOTING_REWARDS_ADDRESS_PARAMS_STEP_RESPOND);
+#	undef   CASE
+		default:
+			THROW(ERR_NOT_IMPLEMENTED);
+		}
+
+		signTxCatalystRegistration_handleVotingRewardsAddress_addressParams_ui_runStep();
+	}
+}
+
+// ============================== NONCE ==============================
+
+enum {
+	HANDLE_NONCE_STEP_DISPLAY = 8600,
+	HANDLE_NONCE_STEP_RESPOND,
+	HANDLE_NONCE_STEP_INVALID,
+};
+
+__noinline_due_to_stack__
+static void signTxCatalystRegistration_handleNonce_ui_runStep()
+{
+	TRACE("UI step %d", subctx->ui_step);
+	TRACE_STACK_USAGE();
+	ui_callback_fn_t* this_fn = signTxCatalystRegistration_handleNonce_ui_runStep;
+
+	UI_STEP_BEGIN(subctx->ui_step);
+
+	UI_STEP(HANDLE_NONCE_STEP_DISPLAY) {
+		STATIC_ASSERT(SIZEOF(subctx->stateData.votingPubKey) == CATALYST_VOTING_PUBLIC_KEY_LENGTH, "wrong voting public key size");
+
+		ui_displayUint64Screen(
+		        "Nonce",
+		        subctx->stateData.nonce,
+		        this_fn
+		);
+	}
+	UI_STEP(HANDLE_NONCE_STEP_RESPOND) {
+		respondSuccessEmptyMsg();
+		advanceState();
+	}
+	UI_STEP_END(HANDLE_NONCE_STEP_INVALID);
+}
+
+__noinline_due_to_stack__
+static void signTxCatalystRegistration_handleNonceAPDU(uint8_t* wireDataBuffer, size_t wireDataSize)
+{
+	{
+		// sanity checks
+		CHECK_STATE(STATE_CATALYST_REGISTRATION_NONCE);
+
+		ASSERT(wireDataSize < BUFFER_SIZE_PARANOIA);
+	}
+	{
+		explicit_bzero(&subctx->stateData, SIZEOF(subctx->stateData));
+	}
+	{
+		// parse data
+		TRACE_BUFFER(wireDataBuffer, wireDataSize);
+		VALIDATE(wireDataSize == 8, ERR_INVALID_DATA);
+		subctx->stateData.nonce = u8be_read(wireDataBuffer);
+		TRACE(
+		        "Catalyst registration nonce: %d",
+		        subctx->stateData.nonce
+		);
+	}
+	{
+		auxDataHashBuilder_catalystRegistration_addNonce(auxDataHashBuilder, subctx->stateData.nonce);
+	}
+
+	// TODO - is it worth declaring a policy for this?
```

Comment:
I think it is an advantage to have all security policies in one place. It is definitely easier to deal with a change in security policy if it has an explicit name. Then it is much easier to find all the relevant code (compared to a plain variable assignment) even if meaning of the policy is not entirely clear from its name.

Reply:
done,  I created basic security policies for all Catalyst registration steps